File: python/kiwi/views/wiki.py
from datetime import datetime, timedelta
from random import randint
import math
import re
from urllib.parse import unquote

from flask import Blueprint, render_template, request, redirect, url_for, abort
from diff_match_patch import diff_match_patch
from sqlalchemy.exc import OperationalError
from markdown import markdown
import bleach
from bs4 import BeautifulSoup

from kiwi import db
from kiwi.models.wiki import Page, History
from kiwi.error import TitleDuplicateException, PageLockException


TITLE_REGEX = "[()\[\]\\n\\r*_`/\\\\]"
LINK_REGEX = "\[\[([^()\[\]\n\r*_`/\\\\]*)\]\]"
LINK_REPLACE = "[\\1](/wiki/\\1)"


bp = Blueprint("wiki", __name__)


@bp.route("/search")
def search():
    string = request.args.get("s")
    if not string:
        return render_template("search.html", search="", pages=None)
    string = string.strip()

    current = request.args.get("p")
    try:
        current = int(current)
        if current <= 0:
            current = 1
    except:
        current = 1

    pages = Page.query.filter(Page.text.ilike(f"%{string}%") | Page.title.ilike(f"%{string}%")).all()
    last = math.ceil(len(pages) / 10)
    show = pages[(current - 1) * 10:current * 10]
    return render_template("search.html", search=string, pages=show, current=current, last=last)


@bp.route("/wiki/<title>")
def view(title):
    page = Page.query.filter_by(title=title).first()

    if page is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return render_template("not-exist.html", title=title)
    else:
        return render_template("view.html", page=page)


@bp.route("/wiki")
def random():
    count = Page.query.count()

    if count > 0:
        index = randint(0, count - 1)
        page = Page.query.offset(index).first()
        return redirect(url_for("wiki.view", title=page.title))
    else:
        return redirect(url_for("wiki.view", title="kiwikiwi"))


@bp.route("/back/<title>/<int:event>")
def back(title, event):
    back = make(title, event)

    if back is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return redirect(url_for("wiki.history", title=title))
    else:
        return render_template("back.html", title=title, page=back, event=event)


@bp.route("/edit/<title>", methods=("GET", "POST"))
def edit(title, summary=None):
    page = Page.query.filter_by(title=title).first()

    if request.method == "GET":
        if page is None:
            if re.search(TITLE_REGEX, title):
                abort(404)
            page = Page(title=title, content="")

        return render_template("edit.html", page=page, summary=summary)
    elif request.method == "POST":
        new_title = request.form["title"]
        content = request.form["content"]
        summary = request.form["summary"]

        if re.search(TITLE_REGEX, new_title):
            new_title = replace(new_title, TITLE_REGEX, "")
            page = Page(title=new_title, content=content)
            return render_template("edit.html", page=page, summary=summary)

        try:
            updated = update(title, new_title, content, summary)
        except (TitleDuplicateException, PageLockException) as e:
            page.title = title
            page.content = content
            return render_template("edit.html", page=page, summary=summary)

        return redirect(url_for("wiki.view", title=updated.title))


@bp.route("/history/<title>")
def history(title):
    page = Page.query.filter_by(title=title).first()

    if page is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return redirect(url_for("wiki.view", title=title))
    else:
        last = math.ceil(len(page.historys) / 10)
        historys = page.historys[:10]
        return render_template("history.html", page=page, current=1, last=last, historys=historys)


@bp.route("/history/<title>/<int:current>")
def history_page(title, current):
    page = Page.query.filter_by(title=title).first()

    if page is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return redirect(url_for("wiki.view", title=title))
    else:
        last = math.ceil(len(page.historys) / 10)
        historys = page.historys[(current - 1) * 10:current * 10]
        return render_template("history.html", page=page, current=current, last=last, historys=historys)


@bp.route("/diff/<title>/<int:event>")
def diff(title, event):
    page = Page.query.filter_by(title=title).first()
    if page is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return redirect(url_for("wiki.view", title=title))

    history = History.query.filter_by(event=event, page_id=page.id).first()

    if history is None:
        return redirect(url_for("wiki.history", title=title))

    title_diff = history.title
    title_diff = replace(title_diff, "\n\+([^\n]*)", "\n##ins##+\\1##/ins##")
    title_diff = replace(title_diff, "\n\-([^\n]*)", "\n##del##-\\1##/del##")
    title_diff = replace(title_diff, "@@\s\-\d+,{0,1}\d*\s\+\d+,{0,1}\d*\s@@\n{0,1}", "")
    title_diff = bleach.clean(unquote(title_diff))
    title_diff = replace(title_diff, "##(ins|/ins|del|/del)##", "<\\1>")
    content_diff = history.content
    content_diff = replace(content_diff, "\n\+([^\n]*)", "\n##ins##+\\1##/ins##")
    content_diff = replace(content_diff, "\n\-([^\n]*)", "\n##del##-\\1##/del##")
    content_diff = replace(content_diff, "@@\s\-\d+,{0,1}\d*\s\+\d+,{0,1}\d*\s@@\n{0,1}", "")
    content_diff = replace(content_diff, "(%0D)*%0A", "%0A ")
    content_diff = bleach.clean(unquote(content_diff))
    content_diff = replace(content_diff, "##(ins|/ins|del|/del)##", "<\\1>")

    return render_template("diff.html", title=title, page=page, history=history, diffs=(title_diff, content_diff))

def replace(string, find, replace):
    return re.sub(pattern=find, repl=replace, string=string)


@bp.route("/rehash/<title>/<int:event>")
def rehash(title, event):
    back = make(title, event)

    if back is None:
        if re.search(TITLE_REGEX, title):
            abort(404)
        return redirect(url_for("wiki.view", title=title))

    try:
        updated = update(title, back.title, back.content, "rehash(" + str(event) + ")")
    except (TitleDuplicateException, PageLockException) as e:
        return redirect(url_for("wiki.back", title=title, event=event))

    return redirect(url_for("wiki.view", title=updated.title))


def make(title, event):
    page = Page.query.filter_by(title=title).first()
    if page is None:
        return None

    history = History.query.filter_by(event=event, page_id=page.id).first()

    if page is None or history is None:
        return None

    back_title = ""
    back_content = ""

    for history in reversed(page.historys):
        if history.event > event:
            break
        back_title = apply_patch(back_title, history.title)
        back_content = apply_patch(back_content, history.content)

    back_content_linked = replace(back_content, LINK_REGEX, LINK_REPLACE)
    back_html = markdown(bleach.clean(back_content_linked))

    return Page(title=back_title, content=back_content, html=back_html)


def update(title, new_title, content, summary):
    page = Page.query.filter_by(title=title).first()

    if title != new_title:
        new_page = Page.query.filter_by(title=new_title).first()

        if new_page is not None:
            raise TitleDuplicateException("page with new title already exists")

    if page is None:
        content_linked = replace(content, LINK_REGEX, LINK_REPLACE)
        html = markdown(bleach.clean(content_linked))
        text = BeautifulSoup(html, features="html.parser").get_text()
        page = Page(title=new_title, content=content, html=html, text=text)
        db.session.add(page)

        title_patch = get_patch("", new_title)
        content_patch = get_patch("", content)

        history = History(page=page, summary=summary or "create", title=title_patch, content=content_patch)
        db.session.add(history)

        db.session.commit()

        return page

    if page.lock is not None and page.lock <= datetime.now():
        page.lock_id = None
        page.lock = None
        db.session.commit()

    lock_id = randint(0, 2147483647)
    lock = datetime.now() + timedelta(seconds=60)

    try:
        Page.query.filter_by(id=page.id, lock_id=None).update(dict(lock=lock, lock_id=lock_id))
        db.session.commit()
    except OperationalError as e:
        db.session.rollback()

    locked = Page.query.filter_by(title=title, lock=lock, lock_id=lock_id).first()

    if locked is None:
        raise PageLockException("page is locked")

    title_patch = get_patch(locked.title, new_title)
    content_patch = get_patch(locked.content, content)
    event = locked.historys[0].event + 1
    history = History(page=locked, summary=summary or "edit", title=title_patch, content=content_patch, event=event, write=datetime.now())
    db.session.add(history)

    content_linked = replace(content, LINK_REGEX, LINK_REPLACE)
    html = markdown(bleach.clean(content_linked))
    text = BeautifulSoup(html, features="html.parser").get_text()
    locked.title = new_title
    locked.content = content
    locked.html = html
    locked.text = text
    locked.lock = None
    locked.lock_id = None
    locked.refresh = datetime.now()
    db.session.commit()

    return locked


def get_patch(text1, text2):
    dmp = diff_match_patch()
    diff = dmp.diff_main(text1, text2, False)
    dmp.diff_cleanupSemantic(diff)
    patch = dmp.patch_make(diff)
    return dmp.patch_toText(patch)


def apply_patch(text, patch_text):
    dmp = diff_match_patch()
    patch = dmp.patch_fromText(patch_text)
    new_text, _ = dmp.patch_apply(patch, text)
    return new_text

File: javascript/wiki.js
const {Op} = require("sequelize");

const {Page, History} = require("./models");
const {TitleDuplicateError, PageLockError} = require("./error");
const DiffMatchPatch = require("diff-match-patch");

const TITLE_REGEX = /[\(\)\[\]\n\r*_`/\\]/g;

function index() {
	return {name: "index"};
}

async function random() {
	let count = await Page.count();

	if (count > 0) {
		let index = Math.floor(Math.random() * count);
		let page = await Page.findOne({offset: index});
		return {redirect: `/wiki/${page.title}`};
	} else {
		return {redirect: "/wiki/kiwikiwi"};
	}
}

async function search(string, current) {
	if (!string || string.trim() === "") {
		return {name: "search", data: {string: "", pages: null}};
	}
	string = string.trim();

	current = +current;
	if (!current || current <= 0) {
		current = 1;
	}

	let pages = await Page.findAll({
		where: {
			[Op.or]: [
				{title: {[Op.like]: `%${string}%`}},
				{text: {[Op.like]: `%${string}%`}},
			],
		},
	});
	let last = Math.ceil(pages.length / 10);
	let show = pages.slice((current - 1) * 10, current * 10);
	return {name: "search", data: {string, current, last, pages: show}};
}

async function view(title) {
	const page = await Page.findOne({where: {title}});

	if (page === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		return {name: "not-exist", data: {page: {title}}};
	} else {
		return {name: "view", data: {page}};
	}
}

async function editView(title) {
	const page = await Page.findOne({where: {title}});

	if (page === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		const emptyPage = Page.build({title});
		emptyPage.newTitle = title;
		return {name: "edit", data: {page: emptyPage}};
	} else {
		page.newTitle = title;
		return {name: "edit", data: {page}};
	}
}

async function editEdit(title, newTitle, summary, content) {
	if (newTitle.match(TITLE_REGEX)) {
		newTitle = newTitle.replace(TITLE_REGEX, "");
		let page = Page.build({title, content});
		page.newTitle = newTitle;
		page.summary = summary;
		return {name: "edit", data: {page}};
	}

	let updated;
	try {
		updated = await update(title, newTitle, summary, content);
	} catch (err) {
		if (err instanceof TitleDuplicateError || err instanceof PageLockError) {
			let page = Page.build({title, content});
			page.newTitle = title;
			return {name: "edit", data: {page}};
		} else {
			throw err;
		}
	}

	return {redirect: `/wiki/${updated.title}`};
}

async function history(title, current = 1) {
	current = +current;
	if (!current) {
		current = 1;
	}

	let page = await Page.findOne({
		where: {title},
		include: [History],
		order: [[History, "event", "DESC"]],
	});

	if (page === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		return {redirect: `/wiki/${title}`};
	} else {
		last = Math.ceil(page.histories.length / 10);
		if (current < 1) current = 1;
		else if (current > last) current = last;

		page.histories = page.histories.slice((current - 1) * 10, current * 10);
		page.current = current;
		page.last = last;
		return {name: "history", data: {page}};
	}
}

async function back(title, event) {
	let back = await make(title, event);

	if (back === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		return {redirect: `/wiki/history/${title}`};
	} else {
		back.event = event;
		return {name: "back", data: {page: back}};
	}
}

async function diff(title, event) {
	let page = await Page.findOne({where: {title}});

	if (page === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		return {redirect: `/wiki/${title}`};
	}

	let history = await History.findOne({where: {event, pageId: page.id}});

	if (history === null) {
		return {redirect: `/history/${title}`};
	}

	let titleDiff = formatDiff(history.title || "");
	let contentDiff = formatDiff(history.content || "");

	page.event = event;
	page.titleDiff = titleDiff;
	page.contentDiff = contentDiff;
	return {name: "diff", data: {page}};
}

function formatDiff(diff) {
	diff = diff.replace(/\n\+([^\n]*)/g, "\n##ins##+$1##/ins##");
	diff = diff.replace(/\n\-([^\n]*)/g, "\n##del##-$1##/del##");
	diff = diff.replace(/@@\s\-\d+,{0,1}\d*\s\+\d+,{0,1}\d*\s@@\n{0,1}/g, "");
	diff = diff.replace(/(%0D)*%0A/g, "%0A ");
	diff = decodeURI(diff);
	diff = diff.replace(/&/g, "&amp;")
		.replace(/</g, "&lt;")
		.replace(/>/g, "&gt;")
		.replace(/"/g, "&quot;")
		.replace(/'/g, "&#39;");
	return diff.replace(/##(ins|\/ins|del|\/del)##/g, "<$1>");
}

async function rehash(title, event) {
	let back = await make(title, event);

	if (back === null) {
		if (title.match(TITLE_REGEX)) {
			return null;
		}
		return {redirect: `/wiki/${title}`};
	}

	try {
		updated = await update(title, back.title, `rehash(${event})`, back.content);
	} catch (err) {
		if (err instanceof TitleDuplicateError || err instanceof PageLockError) {
			return {redirect: `/back/${title}/${event}`};
		} else {
			throw err;
		}
	}

	return {redirect: `/wiki/${updated.title}`};
}

async function make(title, event) {
	let page = await Page.findOne({
		where: {title},
		include: [History],
		order: [[History, "event", "ASC"]],
	});

	if (page === null) {
		return null;
	}

	let history = await History.findOne({where: {event, pageId: page.id}});

	if (history === null) {
		return null;
	}

	backTitle = "";
	backContent = "";

	for (let history of page.histories) {
		if (history.event > event) break;
		backTitle = applyPatch(backTitle, history.title);
		backContent = applyPatch(backContent, history.content);
	}

	return Page.build({title: backTitle, content: backContent});
}

async function update(title, newTitle, summary, content) {
	let page = await Page.findOne({where: {title}});

	if (title !== newTitle) {
		let newPage = await Page.findOne({where: {title: newTitle}});

		if (newPage !== null) {
			throw new TitleDuplicateError("page with new title already exists");
		}
	}

	if (page === null) {
		page = Page.build({title: newTitle, content});
		await page.save();

		let titlePatch = getPatch("", newTitle);
		let contentPatch = getPatch("", content);

		let history = History.build({
			summary: summary || "create",
			title: titlePatch,
			content: contentPatch,
		});
		await history.save()
			.then(() => history.setPage(page));

		return page;
	}

	if (page.lock !== null && page.lock <= new Date()) {
		page.lockId = null;
		page.lock = null;
		await page.save();
	}

	let id = page.id;
	let lockId = Math.floor(Math.random() * 2147483647);
	let lock = new Date(+new Date() + 60000);

	await Page.update(
		{lockId, lock},
		{where: {
			id,
			lockId: null,
		}},
	);

	let locked = await Page.findOne({
		where :{
			title,
			lock,
			lockId,
		},
		include: [History],
		order: [[History, "event", "DESC"]],
	});

	if (locked === null) {
		throw new PageLockError("page is locked");
	}

	let titlePatch = getPatch(locked.title, newTitle);
	let contentPatch = getPatch(locked.content, content);
	let event = locked.histories[0].event + 1;
	let history = History.build({
		summary: summary || "edit",
		event,
		title: titlePatch,
		content: contentPatch,
	});
	await history.save()
		.then(() => history.setPage(locked));

	locked.title = newTitle;
	locked.content = content;
	locked.lockId = null;
	locked.lock = null;
	locked.refresh = new Date();
	await locked.save();

	return locked;
}

function getPatch(text1, text2) {
	let dmp = new DiffMatchPatch();
	let diff = dmp.diff_main(text1, text2, false);
	dmp.diff_cleanupSemantic(diff);
	let patch = dmp.patch_make(diff);
	return dmp.patch_toText(patch);
}

function applyPatch(text, patchText) {
	let dmp = new DiffMatchPatch();
	let patch = dmp.patch_fromText(patchText);
	let newText = dmp.patch_apply(patch, text);
	return newText[0];
}

module.exports = {
	index,
	random,
	search,
	view,
	editView, editEdit,
	history, back, diff, rehash,
}
